### DISCUSSION TITLE
Using SHAP on OneHotEncoded data - does it generate non-sensical data?

### DISCUSSION THREAD

**Discussion by ra0405** [shap/shap#3221] [Q&A]
I've seen that there can be issues when using LIME on OneHotEncoded data as the perturbations can result in non-sensical data - e.g. 0 for male feature and 0 for female feature. Therefore when using LIME you need to remove the OHE step.

Does the same issue exist for SHAP?

**Comment by detrin** [comment]
I don't understand how you define nonsensical. Please have a look at https://christophm.github.io/interpretable-ml-book/shapley.html to understand how shapley values work

**Comment by znacer** [comment]
Unless you are using TreeSHAP, you provide background data to compute SHAP values.
In the case of TreeSHAP, the algorithm follow paths. Therefore, if a non sensical data occurs, it means this concerned features are not affected by the tree and therefore would have 0 attribution.
In other cases, as long as these data do not occurs in the background you provided, the situation won't happend.

An other way to avoid this, when using agnostic explainers, is by wrapping you model and do the one-hot encoding in the wrapper so that this situation just cannot happened :
```py
def forward_wrapper(x):
    one_hot_x = one_hot_preprocessing(x)
    return model.predict_proba(one_hot_x)
```
and then explain the wrapped function.
